.github/workflows/docker-dev.yaml

```diff
@@ -0,0 +1,66 @@
+name: Build and push development images
```

Comment:
The repo these jobs push to is for public consumption, and discussion with Harry raised what I think is a legitimate point: having a proliferation of tags there is confusing. Users shouldn't have any reason to pull individual commit releases. They should almost always be running release builds (builds for Git tags) and may want to test pre-release builds (e.g. we solve a bug, merge the fix, and ask them to try the latest main/next tag to confirm the fix without having to build and push their own image somewhere).

Test CI jobs have a different set of needs, as they're interested in individual commits on any branch, and almost all of the artifacts (anything that doesn't end up being on the tip of main or next) aren't of interest outside their specific test run. As such, I like the approach taken in the Railgun test (just build and run the Go binary) since the jobs clean up after themselves. Generating and uploading image artifacts would be more appropriate if we needed to trigger tests outside the controller repo, but as of yet we don't.

I recognize the interest in having the same artifact regardless, but divergence there hasn't been a source of issues for us. There's not really anything of significance in the images beyond the Go binary: they're basically just necessary packaging to provide something K8S can run that includes version metadata.

The current split between the jobs is a simple means of handling the different sources for the Docker tag names, so that they're identical aside from the `on` criteria and output in the `extract_shortref` step and still simple and straightforward. Unifying them is likely possible, but requires a lot more complex conditional logic and manipulation of step outputs to produce the final tag list. To Shane's point earlier, this is a start, and we can expand on it later.

Reply:
I've had good experiences with https://github.com/marketplace/actions/docker-meta

TLDR 
![image](https://user-images.githubusercontent.com/697188/114482493-aba4ac00-9bd4-11eb-84fd-9be29b831fc8.png)